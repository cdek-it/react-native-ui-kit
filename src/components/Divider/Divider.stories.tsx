import type { Meta, StoryObj } from '@storybook/react'
import { IconCircleCheck, IconInfoCircle } from '@tabler/icons-react-native'

import { View } from 'react-native'

import { Divider } from './Divider'

const icons = { IconInfoCircle, IconCircleCheck, none: undefined }

const meta: Meta<typeof Divider> = {
  title: 'Divider',
  component: Divider,
  decorators: (Story) => (
    <View style={{ height: 300 }}>
      <Story />
    </View>
  ),
  args: {
    align: 'center',
    layout: 'horizontal',
    showContent: true,
    showIcon: true,
    text: 'subtitle sm',
    type: 'solid',
    Icon: IconCircleCheck,
  },
  argTypes: {
    align: { control: 'radio', options: ['start', 'center', 'end'] },
    layout: { control: 'radio', options: ['horizontal', 'vertical'] },
    type: { control: 'radio', options: ['solid', 'dash'] },
    Icon: { control: 'radio', options: Object.keys(icons), mapping: icons },
  },
}

export default meta

type Story = StoryObj<typeof Divider>

const DividerStory: Story = {}

export { DividerStory as Divider }
