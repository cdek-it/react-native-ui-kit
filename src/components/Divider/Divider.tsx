import type { Icon } from '@tabler/icons-react-native'
import { memo, useMemo } from 'react'
import { View, type ViewStyle, type StyleProp } from 'react-native'

import { makeStyles } from '../../utils/makeStyles'
import { Subtitle } from '../Typography'

export interface DividerProps {
  /**
   * Выбор расположения контента относительно линии.
   * В фигме вместо `start/end` используются `left/right`
   * для горизонтальной ориентации и `top/bottom` для вертикальной
   * @default 'center'
   */
  align?: 'start' | 'center' | 'end'

  /**
   * Выбор ориентации компонента (горизонтальная либо вертикальная)
   * @default 'horizontal'
   */
  layout?: 'horizontal' | 'vertical'

  /**
   * Показать или скрыть контейнер контента поверх линии
   * @default true
   */
  showContent?: boolean

  /**
   * Показать или скрыть иконку внутри контейнера с контентом
   * @default true
   */
  showIcon?: boolean

  /** Текст */
  text?: string

  /**
   * Выбор стиля линии
   * @default 'solid'
   */
  type?: 'solid' | 'dash'

  /** Дополнительная стилизация для контейнера компонента */
  style?: StyleProp<ViewStyle>

  /** Иконка из библиотеки Tabler */
  Icon?: Icon
}

/**
 * Используется для визуального разделения контента
 * @see https://www.figma.com/design/4TYeki0MDLhfPGJstbIicf/UI-kit-PrimeFace-(DS)?node-id=484-5178
 */
export const Divider = memo<DividerProps>(
  ({
    align = 'center',
    layout = 'horizontal',
    showContent: showContentProp = true,
    showIcon: showIconProp = true,
    text,
    type = 'solid',
    style,
    Icon,
  }) => {
    const styles = useStyles()
    const isVertical = useMemo(() => layout === 'vertical', [layout])
    const showIcon = useMemo(
      () => !!(showIconProp && Icon),
      [Icon, showIconProp]
    )

    const showContent = useMemo(
      () => !!(showContentProp && (showIcon || text)),
      [showContentProp, showIcon, text]
    )

    const lineStyle = useMemo(
      () => [styles.line, type === 'dash' && styles.lineDash],
      [styles.line, styles.lineDash, type]
    )

    return (
      <View
        style={[
          styles.container,
          isVertical && styles.containerVertical,
          showContent &&
            align === 'end' &&
            (isVertical
              ? styles.containerColumnReverse
              : styles.containerRowReverse),
          style,
        ]}
      >
        {showContent ? (
          <>
            <View
              style={[
                styles.lineContainer,
                align !== 'center' && styles.lineContainerSide,
              ]}
            >
              <View style={lineStyle} />
            </View>

            <View
              style={[styles.content, isVertical && styles.contentVertical]}
            >
              {showIcon && Icon ? (
                <Icon
                  height={styles.icon.height}
                  style={styles.icon}
                  width={styles.icon.width}
                />
              ) : null}
              {text ? (
                <Subtitle color='secondary' style={styles.text}>
                  {text}
                </Subtitle>
              ) : null}
            </View>
          </>
        ) : null}

        <View style={styles.lineContainer}>
          <View style={lineStyle} />
        </View>
      </View>
    )
  }
)

const useStyles = makeStyles(({ spacing, theme, sizing, typography }) => ({
  container: {
    minHeight: sizing.Height['h-2'],
    flexDirection: 'row',
    justifyContent: 'center',
    alignItems: 'center',
  },
  containerVertical: {
    minWidth: sizing.Width['w-2'],
    flexDirection: 'column',
    flexGrow: 1,
    alignSelf: 'flex-start',
  },
  containerRowReverse: { flexDirection: 'row-reverse' },
  containerColumnReverse: { flexDirection: 'column-reverse' },
  lineContainer: {
    flexGrow: 1,
    flexBasis: spacing.Gap['gap-4'],
    width: 1,
    height: 1,
    overflow: 'hidden',
  },
  lineContainerSide: { flexGrow: 0 },
  line: {
    width: '100%',
    height: '100%',
    borderWidth: 1,
    borderColor: theme.General.dividerColor,
  },
  lineDash: { borderStyle: 'dashed' },
  content: {
    flexShrink: 1,
    flexDirection: 'row',
    alignItems: 'center',
    paddingHorizontal: theme.General.inlineSpacing,
    gap: spacing.Gap['gap-2'],
    backgroundColor: theme.Panel.Panel.Body.panelContentBg,
  },
  contentVertical: {
    paddingHorizontal: 0,
    paddingVertical: theme.General.inlineSpacing,
  },
  icon: {
    width: typography.Size['text-base'],
    height: typography.Size['text-base'],
    color: theme.General.textSecondaryColor,
  },
  text: { flexShrink: 1 },
}))
