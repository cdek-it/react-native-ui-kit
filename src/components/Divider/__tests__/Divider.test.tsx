import { IconCheck } from '@tabler/icons-react-native'
import { render } from '@testing-library/react-native'

import { Divider, type DividerProps } from '../Divider'

describe('Divider', () => {
  const snapshotCases: Array<[DividerProps]> = [
    [{}],
    [
      {
        Icon: IconCheck,
        align: 'start',
        layout: 'horizontal',
        style: { margin: 10 },
      },
    ],
    [
      {
        Icon: IconCheck,
        align: 'end',
        layout: 'vertical',
        text: 'test',
        type: 'solid',
      },
    ],
    [
      {
        Icon: IconCheck,
        align: 'end',
        layout: 'vertical',
        showIcon: false,
        type: 'dash',
      },
    ],
    [
      {
        Icon: IconCheck,
        align: 'end',
        layout: 'horizontal',
        text: 'test',
        showIcon: false,
      },
    ],
    [
      {
        Icon: IconCheck,
        align: 'center',
        layout: 'vertical',
        text: 'test',
        showContent: false,
      },
    ],
    [{ align: 'end', layout: 'horizontal', text: 'test', showContent: false }],
  ]

  test.each(snapshotCases)('%p', (props) => {
    const { toJSON } = render(<Divider {...props} />)

    expect(toJSON()).toMatchSnapshot()
  })
})
