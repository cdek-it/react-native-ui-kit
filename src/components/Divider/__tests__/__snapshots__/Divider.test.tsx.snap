// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Divider {"Icon": [Object], "align": "center", "layout": "vertical", "showContent": false, "text": "test"} 1`] = `
<View
  style={
    [
      {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "center",
        "minHeight": 28,
      },
      {
        "alignSelf": "flex-start",
        "flexDirection": "column",
        "flexGrow": 1,
        "minWidth": 28,
      },
      false,
      undefined,
    ]
  }
>
  <View
    style={
      {
        "flexBasis": 14,
        "flexGrow": 1,
        "height": 1,
        "overflow": "hidden",
        "width": 1,
      }
    }
  >
    <View
      style={
        [
          {
            "borderColor": "#dbdada",
            "borderWidth": 1,
            "height": "100%",
            "width": "100%",
          },
          false,
        ]
      }
    />
  </View>
</View>
`;

exports[`Divider {"Icon": [Object], "align": "end", "layout": "horizontal", "showIcon": false, "text": "test"} 1`] = `
<View
  style={
    [
      {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "center",
        "minHeight": 28,
      },
      false,
      {
        "flexDirection": "row-reverse",
      },
      undefined,
    ]
  }
>
  <View
    style={
      [
        {
          "flexBasis": 14,
          "flexGrow": 1,
          "height": 1,
          "overflow": "hidden",
          "width": 1,
        },
        {
          "flexGrow": 0,
        },
      ]
    }
  >
    <View
      style={
        [
          {
            "borderColor": "#dbdada",
            "borderWidth": 1,
            "height": "100%",
            "width": "100%",
          },
          false,
        ]
      }
    />
  </View>
  <View
    style={
      [
        {
          "alignItems": "center",
          "backgroundColor": "#ffffff",
          "flexDirection": "row",
          "flexShrink": 1,
          "gap": 7,
          "paddingHorizontal": 7,
        },
        false,
      ]
    }
  >
    <Text
      style={
        [
          {
            "fontSize": 12.25,
            "fontWeight": 700,
            "textTransform": "uppercase",
          },
          {
            "color": "rgba(0, 0, 0, 0.6000)",
          },
          false,
          {
            "flexShrink": 1,
          },
        ]
      }
    >
      test
    </Text>
  </View>
  <View
    style={
      {
        "flexBasis": 14,
        "flexGrow": 1,
        "height": 1,
        "overflow": "hidden",
        "width": 1,
      }
    }
  >
    <View
      style={
        [
          {
            "borderColor": "#dbdada",
            "borderWidth": 1,
            "height": "100%",
            "width": "100%",
          },
          false,
        ]
      }
    />
  </View>
</View>
`;

exports[`Divider {"Icon": [Object], "align": "end", "layout": "vertical", "showIcon": false, "type": "dash"} 1`] = `
<View
  style={
    [
      {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "center",
        "minHeight": 28,
      },
      {
        "alignSelf": "flex-start",
        "flexDirection": "column",
        "flexGrow": 1,
        "minWidth": 28,
      },
      false,
      undefined,
    ]
  }
>
  <View
    style={
      {
        "flexBasis": 14,
        "flexGrow": 1,
        "height": 1,
        "overflow": "hidden",
        "width": 1,
      }
    }
  >
    <View
      style={
        [
          {
            "borderColor": "#dbdada",
            "borderWidth": 1,
            "height": "100%",
            "width": "100%",
          },
          {
            "borderStyle": "dashed",
          },
        ]
      }
    />
  </View>
</View>
`;

exports[`Divider {"Icon": [Object], "align": "end", "layout": "vertical", "text": "test", "type": "solid"} 1`] = `
<View
  style={
    [
      {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "center",
        "minHeight": 28,
      },
      {
        "alignSelf": "flex-start",
        "flexDirection": "column",
        "flexGrow": 1,
        "minWidth": 28,
      },
      {
        "flexDirection": "column-reverse",
      },
      undefined,
    ]
  }
>
  <View
    style={
      [
        {
          "flexBasis": 14,
          "flexGrow": 1,
          "height": 1,
          "overflow": "hidden",
          "width": 1,
        },
        {
          "flexGrow": 0,
        },
      ]
    }
  >
    <View
      style={
        [
          {
            "borderColor": "#dbdada",
            "borderWidth": 1,
            "height": "100%",
            "width": "100%",
          },
          false,
        ]
      }
    />
  </View>
  <View
    style={
      [
        {
          "alignItems": "center",
          "backgroundColor": "#ffffff",
          "flexDirection": "row",
          "flexShrink": 1,
          "gap": 7,
          "paddingHorizontal": 7,
        },
        {
          "paddingHorizontal": 0,
          "paddingVertical": 7,
        },
      ]
    }
  >
    <RNSVGSvgView
      align="xMidYMid"
      bbHeight={14}
      bbWidth={14}
      fill="none"
      focusable={false}
      height={14}
      meetOrSlice={0}
      minX={0}
      minY={0}
      stroke="currentColor"
      strokeLinecap="round"
      strokeLinejoin="round"
      strokeWidth={2}
      style={
        [
          {
            "backgroundColor": "transparent",
            "borderWidth": 0,
          },
          {
            "color": "rgba(0, 0, 0, 0.6000)",
            "height": 14,
            "width": 14,
          },
          {
            "flex": 0,
            "height": 14,
            "width": 14,
          },
        ]
      }
      vbHeight={24}
      vbWidth={24}
      width={14}
      xmlns="http://www.w3.org/2000/svg"
    >
      <RNSVGGroup
        color="rgba(0, 0, 0, 0.6000)"
        fill={null}
        propList={
          [
            "fill",
            "stroke",
            "strokeWidth",
            "strokeLinecap",
            "strokeLinejoin",
          ]
        }
        stroke={
          {
            "type": 2,
          }
        }
        strokeLinecap={1}
        strokeLinejoin={1}
        strokeWidth={2}
      >
        <RNSVGPath
          color="rgba(0, 0, 0, 0.6000)"
          d="M5 12l5 5l10 -10"
          fill={null}
          propList={
            [
              "fill",
              "stroke",
              "strokeWidth",
              "strokeLinecap",
              "strokeLinejoin",
            ]
          }
          stroke={
            {
              "type": 2,
            }
          }
          strokeLinecap={1}
          strokeLinejoin={1}
          strokeWidth={2}
        />
      </RNSVGGroup>
    </RNSVGSvgView>
    <Text
      style={
        [
          {
            "fontSize": 12.25,
            "fontWeight": 700,
            "textTransform": "uppercase",
          },
          {
            "color": "rgba(0, 0, 0, 0.6000)",
          },
          false,
          {
            "flexShrink": 1,
          },
        ]
      }
    >
      test
    </Text>
  </View>
  <View
    style={
      {
        "flexBasis": 14,
        "flexGrow": 1,
        "height": 1,
        "overflow": "hidden",
        "width": 1,
      }
    }
  >
    <View
      style={
        [
          {
            "borderColor": "#dbdada",
            "borderWidth": 1,
            "height": "100%",
            "width": "100%",
          },
          false,
        ]
      }
    />
  </View>
</View>
`;

exports[`Divider {"Icon": [Object], "align": "start", "layout": "horizontal", "style": [Object]} 1`] = `
<View
  style={
    [
      {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "center",
        "minHeight": 28,
      },
      false,
      false,
      {
        "margin": 10,
      },
    ]
  }
>
  <View
    style={
      [
        {
          "flexBasis": 14,
          "flexGrow": 1,
          "height": 1,
          "overflow": "hidden",
          "width": 1,
        },
        {
          "flexGrow": 0,
        },
      ]
    }
  >
    <View
      style={
        [
          {
            "borderColor": "#dbdada",
            "borderWidth": 1,
            "height": "100%",
            "width": "100%",
          },
          false,
        ]
      }
    />
  </View>
  <View
    style={
      [
        {
          "alignItems": "center",
          "backgroundColor": "#ffffff",
          "flexDirection": "row",
          "flexShrink": 1,
          "gap": 7,
          "paddingHorizontal": 7,
        },
        false,
      ]
    }
  >
    <RNSVGSvgView
      align="xMidYMid"
      bbHeight={14}
      bbWidth={14}
      fill="none"
      focusable={false}
      height={14}
      meetOrSlice={0}
      minX={0}
      minY={0}
      stroke="currentColor"
      strokeLinecap="round"
      strokeLinejoin="round"
      strokeWidth={2}
      style={
        [
          {
            "backgroundColor": "transparent",
            "borderWidth": 0,
          },
          {
            "color": "rgba(0, 0, 0, 0.6000)",
            "height": 14,
            "width": 14,
          },
          {
            "flex": 0,
            "height": 14,
            "width": 14,
          },
        ]
      }
      vbHeight={24}
      vbWidth={24}
      width={14}
      xmlns="http://www.w3.org/2000/svg"
    >
      <RNSVGGroup
        color="rgba(0, 0, 0, 0.6000)"
        fill={null}
        propList={
          [
            "fill",
            "stroke",
            "strokeWidth",
            "strokeLinecap",
            "strokeLinejoin",
          ]
        }
        stroke={
          {
            "type": 2,
          }
        }
        strokeLinecap={1}
        strokeLinejoin={1}
        strokeWidth={2}
      >
        <RNSVGPath
          color="rgba(0, 0, 0, 0.6000)"
          d="M5 12l5 5l10 -10"
          fill={null}
          propList={
            [
              "fill",
              "stroke",
              "strokeWidth",
              "strokeLinecap",
              "strokeLinejoin",
            ]
          }
          stroke={
            {
              "type": 2,
            }
          }
          strokeLinecap={1}
          strokeLinejoin={1}
          strokeWidth={2}
        />
      </RNSVGGroup>
    </RNSVGSvgView>
  </View>
  <View
    style={
      {
        "flexBasis": 14,
        "flexGrow": 1,
        "height": 1,
        "overflow": "hidden",
        "width": 1,
      }
    }
  >
    <View
      style={
        [
          {
            "borderColor": "#dbdada",
            "borderWidth": 1,
            "height": "100%",
            "width": "100%",
          },
          false,
        ]
      }
    />
  </View>
</View>
`;

exports[`Divider {"align": "end", "layout": "horizontal", "showContent": false, "text": "test"} 1`] = `
<View
  style={
    [
      {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "center",
        "minHeight": 28,
      },
      false,
      false,
      undefined,
    ]
  }
>
  <View
    style={
      {
        "flexBasis": 14,
        "flexGrow": 1,
        "height": 1,
        "overflow": "hidden",
        "width": 1,
      }
    }
  >
    <View
      style={
        [
          {
            "borderColor": "#dbdada",
            "borderWidth": 1,
            "height": "100%",
            "width": "100%",
          },
          false,
        ]
      }
    />
  </View>
</View>
`;

exports[`Divider {} 1`] = `
<View
  style={
    [
      {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "center",
        "minHeight": 28,
      },
      false,
      false,
      undefined,
    ]
  }
>
  <View
    style={
      {
        "flexBasis": 14,
        "flexGrow": 1,
        "height": 1,
        "overflow": "hidden",
        "width": 1,
      }
    }
  >
    <View
      style={
        [
          {
            "borderColor": "#dbdada",
            "borderWidth": 1,
            "height": "100%",
            "width": "100%",
          },
          false,
        ]
      }
    />
  </View>
</View>
`;
